src/workerd/api/sockets.c++

```diff
@@ -60,122 +69,23 @@ jsg::Ref<Socket> connectImpl(
   return connectImplNoOutputLock(js, kj::mv(actualFetcher), kj::mv(address));
 }
 
-InitData initialiseSocket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) {
-  auto& context = IoContext::current();
-
-  // Initialise the readable/writable streams with a custom AsyncIoStream that waits for the
-  // completion of `connectionPromise` before performing reads/writes.
-  auto stream = kj::refcounted<PipelinedAsyncIoStream>(kj::mv(connectionPromise));
-  auto sysStreams = newSystemMultiStream(kj::addRef(*stream), StreamEncoding::IDENTITY, context);
-
-  return {
-    .readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable)),
-    .writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable)),
-    .closeFulfiller = IoContext::current().addObject(
-        kj::heap<kj::PromiseFulfillerPair<void>>(kj::newPromiseAndFulfiller<void>()))
-  };
-}
-
-Socket::Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) :
-    Socket(initialiseSocket(kj::mv(connectionPromise))) {};
-
-jsg::Promise<void> Socket::close(jsg::Lock& js) {
-  if (!closeFulfiller->fulfiller->isWaiting()) {
-    return js.resolvedPromise();
+void Socket::close() {
+  // TODO(later): Write test for calling `close` twice in a row and determine whether a "closing"
+  // state is needed.
+  if (isClosed) {
```

Comment:
This `isClosed` check looks race-y since `isClosed` doesn't actually become true until after the async `cancel()` and `abort()` calls below complete. So if `close()` is called twice in a row, both calls will continue on to the `cancel()` and `abort()` below. If the check isn't really important than I'd remove it, but if it is important then it currently doesn't look like it works.

Reply:
Yeah, that's why I have the "todo" in there :)

This check will still help if `close` is called once the `closed` promise resolves by accident for example.

I'd rather keep it and iterate on it to resolve the race later./

Reply:
I would add a "closing" state.

Reply:
Turns out cancelling/aborting the streams multiple times doesn't create any problems, so I ended up just removing these states completely. Upstream I've added a test to ensure this case is covered as well.